docs/cookbook/web_scraping.md

```diff
@@ -0,0 +1,251 @@
+
+```python
+from openai.types.chat import ChatCompletionMessageParam
+from pydantic import BaseModel
+
+from mirascope.core import openai, prompt_template
+
+class WebAssistant(BaseModel):
+    history: list[ChatCompletionMessageParam] = []
+
+    @openai.call(model="gpt-4o")
+    @prompt_template(
+        """
+        SYSTEM:
+        You are an expert web searcher. Your task is to answer the user's question using the provided tools.
+        Use the tool `web_search` once to gather the required information.
+        After using the tool, provide a comprehensive answer to the user's question based on the information you retrieved.
+
+        MESSAGES:
+        {self.history}
+
+        USER:
+        {question}
+        """
+    )
+    def _step(self, question: str): ...
+```
+
+The `WebAssistant` will first check the web for relevant search results, before digging deeper into each web page for information to help answer the user's question.
+Note that we have not defined the `web_search` tool yet, so let's do that next.
+
+## Add DuckDuckGo Tool
+
+Now that we have created our prompt, let’s go ahead and add our web search tool:
+
+```python
+import inspect
+
+from duckduckgo_search import DDGS
+
+def web_search(text: str) -> str:
+    """Search the web for the given text.
+
+    Args:
+        text: The text to search for.
+
+    Returns:
+        Parsed paragraphs of each of the webpages, separated by newlines.
+    """
+    try:
+        results = DDGS(proxy=None).text(text, max_results=5)
+        return "\n\n".join([parse_webpage(result["href"]) for result in results])
+    except Exception as e:
+        return f"{type(e)}: Failed to search the web for text"
+```
+
+We are grabbing the first 5 results that best match our user query and retriving their URLs, for parsing:
+
+```python
+def parse_webpage(self, link: str) -> str:
+    """Parse the paragraphs of the webpage found at `link`.
+
+    Args:
+        link: The URL of the webpage.
+
+    Returns:
+        The parsed paragraphs of the webpage, separated by newlines.
+    """
+    try:
+        response = requests.get(link)
+        soup = BeautifulSoup(response.content, "html.parser")
+        return "\n".join([p.text for p in soup.find_all("p")])
+    except Exception as e:
+        return f"{type(e)}: Failed to parse content from URL"
+```
+
+We use BeautifulSoup to assist in extracting all paragraph tags in the HTML. Now that our tool is setup, we can proceed to implement the Q&A functionality of our Agent.
+
+## Add Q&A Functionality
+
+```python
+import requests
+from bs4 import BeautifulSoup
+from duckduckgo_search import DDGS
+from openai.types.chat import ChatCompletionMessageParam
+from pydantic import BaseModel
+
+from mirascope.core import openai, prompt_template
+
+
+def web_search(text: str) -> str:
```

Comment:
it's weird to me that we duplicate this.

I would rather first just implement the tools, then implement the prompt with the tool (like we do for the blog writing agent)